### DISCUSSION TITLE
Dependency Injection in WPF

### DISCUSSION THREAD

**Discussion by DavidM29** [dotnet/wpf#9022] [Q&A] [answered]
Hello,

I am relatively new to wpf and .net and I am developing a small application.
I know that injecting pages and viewmodels with dependency injection is quite a popular approach, however I was interested if it is also a good approach so specify viewmodels in the xaml of the respective views like this:
`<Page.DataContext>
    <viewModels:MainViewModel x:Name="ViewModel"/>
</Page.DataContext>`

The problem I encountered with this, is that now I am unable to inject services into the viewmodels themselves, however I can still register pages as services, inject dependencies into the pages and then ,through code-behind, manually set properties for these services in the viewmodel.

Basically what I am saying is I want to register only the pages as services and specify the viewmodels in xaml, and in case I need to inject some services into the viewmodel I will do it manually through the code-behind.

Would this be a good approach or should I stick with registering both view and viewmodels as services.

Thanks in advance.

**Comment by lindexi** [comment]
In fact, you can add the ViewModel to DataContext in your xaml.cs code.

**Reply by DavidM29** [reply]
I know, that is what I am doing at the moment, but I would prefer to specify the viewmodel directly in the xaml, because that way autocompletion works correctly and I do not have to write everything by hand, also I can easily see if I have mistake somewhere in the binding. I am interested if using this method has any downsides compared to registering both views and viewmodels as services.

**Comment by jamesnet214** [comment, marked as answer]
In the WPF and .NET environment, there are several approaches to managing pages and view models through dependency injection. One of the most popular methods involves using a framework like Prism. With Prism, when a view is created, the associated view model is automatically generated, and necessary dependencies are injected through the constructor. This allows for efficient and flexible dependency management without explicitly declaring the view model within the XAML.

However, even without using a library like Prism, it is possible to use reflection to identify what instances are required by the constructor of a view model and inject these dependencies directly. This method enables you to benefit from dependency injection even without a framework.

The approach you suggested, registering only the pages as services and specifying the view models in XAML, followed by manually injecting required services through code-behind, is also feasible. This strategy offers flexibility in managing dependencies while lowering the coupling between views and view models. However, it might somewhat limit the reusability and ease of testing of the view models. Therefore, it's important to choose the right approach based on the project's requirements and the team's preferences.

Overall, it is recommended to actively utilize dependency injection and to thoroughly understand and apply the architectural design of proven frameworks for effective software development. This approach significantly contributes to improving the maintainability and scalability of the code.

In summary, actively using dependency injection and understanding and applying the architectural designs of established frameworks, such as Prism, is recommended for effective software development. Prism, in particular, facilitates the integration and enhancement of these practices. Using Prism simplifies the management of dependency injection and view and view model coordination, which substantially aids in enhancing the maintainability and scalability of code. Employing such frameworks can help structure software architecture in a robust and manageable way, making the development process more efficient and effective.
